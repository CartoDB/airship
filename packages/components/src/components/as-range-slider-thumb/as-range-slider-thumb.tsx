import { Component, Element, Event, EventEmitter, Listen, Prop } from '@stencil/core';

@Component({
  shadow: false,
  styleUrl: './as-range-slider-thumb.scss',
  tag: 'as-range-slider-thumb',
})
export class RangeSliderThumb {
  @Prop() public maxValue: number;
  @Prop() public minValue: number;
  @Prop() public value: number;
  @Prop() public formatValue: (value: number) => void;

  @Event() public valueUpdate: EventEmitter<number>;
  @Event() public changeStart: EventEmitter<number | number[]>;
  @Event() public changeEnd: EventEmitter<number | number[]>;

  @Element() public thumbElement: HTMLElement;
  public railElement: HTMLElement;

  public render() {
    const thumbStyles = {
      left: `${this._getPercentage(this.value)}%`
    };

    return (
      <div class='as-range-slider__thumb' style={thumbStyles} data-value={this.value}>
          <span class='as-range-slider__value as-caption as-font-medium'>
            {this._getDisplayValue(this.value)}
          </span>
      </div>);
  }

  @Listen('mousedown')
  @Listen('touchstart')
  public onMouseDown(event: MouseEvent) {
    this.railElement = document.querySelector('.as-range-slider__rail');

    const thumb = event.target as HTMLElement;
    thumb.classList.add('as-range-slider__thumb--moving');

    const handleMove = (eventProperties) => this._handleMove(eventProperties);
    const handleRelease = () => this._handleRelease(thumb, { mouseMove: handleMove, mouseDown: handleRelease });

    document.addEventListener('mousemove', handleMove);
    document.addEventListener('touchmove', handleMove);
    document.addEventListener('mouseup', handleRelease);
    document.addEventListener('touchend', handleRelease);

    this.thumbElement.focus();
  }

  private _handleMove(event: MouseEvent) {
    const barPercentage = (event.pageX - this.railElement.offsetLeft) * 100 / this.railElement.offsetWidth;
    const newValue = (barPercentage * (this.maxValue - this.minValue)) / 100;

    if (newValue < 0 || newValue > this.maxValue) {
      return;
    }

    this.valueUpdate.emit(newValue);
  }

  private _handleRelease(
    thumb: HTMLElement,
    listeners: { mouseMove: (eventProperties) => void, mouseDown: () => void}
  ) {
    document.removeEventListener('mousemove', listeners.mouseMove);
    document.removeEventListener('mouseup', listeners.mouseDown);
    thumb.classList.remove('as-range-slider__thumb--moving');
  }

  private _getPercentage(value: number) {
    return ((value - this.minValue) / (this.maxValue - this.minValue)) * 100;
  }

  private _getDisplayValue(value: number) {
    const displayValue = Math.round(value);
    return (this.formatValue && this.formatValue(value)) || displayValue;
  }
}

export interface Thumb {
  value: number;
}
