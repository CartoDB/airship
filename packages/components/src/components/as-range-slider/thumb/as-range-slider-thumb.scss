@import '../../../../../styles/src/core/_dev';

as-range-slider-thumb {
  .as-range-slider__thumb {
    position: absolute;
    z-index: 2;
    width: 12px;
    height: 12px;
    transform: translate3d(-6px, 0, 0);
    transition: transform 200ms ease, opacity 200ms ease;
  }

  .as-range-slider__value {
    position: absolute;
    bottom: -16px;
    left: 50%;
    transform: translate3d(-53%, 0, 0);
    transition: transform 200ms ease;
    pointer-events: none;

    &--disabled {
      color: var(--as-color-type-03, #{$color-type-03});
    }
  }

  .as-range-slider__thumb:hover .as-range-slider__value {
    transform: translate3d(-53%, 4px, 0);
  }

  .as-range-slider__thumb-handle {
    width: 12px;
    height: 12px;
    transition: transform 200ms ease;
    border: 1px solid var(--as-color-primary, #{$color-primary});
    border-radius: 50%;
    background: #FFF;
    cursor: grab;

    &::before {
      content: '';
      position: absolute;
      top: -15px;
      left: -15px;
      width: 30px;
      height: 30px;
    }

    &:hover,
    &.as-range-slider__thumb-handle--moving {
      transform: scale(1.33);
    }

    &--moving {
      cursor: grabbing;
    }
  }

  .as-range-slider__thumb--disabled {
    .as-range-slider__thumb-handle {
      border: 1px solid var(--as-color-ui-03, #{$color-ui-03});
      background: var(--as-color-ui-02, #{$color-ui-02});

      &:focus {
        background: var(--as-color-ui-02, #{$color-ui-02});
      }
    }
  }

  .as-range-slider__thumb:focus {
    outline: none;

    .as-range-slider__thumb-handle {
      background: var(--as-color-primary, #{$color-primary});
    }
  }

  .as-range-slider__thumb + .as-range-slider__thumb {
    &:hover {
      transform: translate3d(-6px, 0, 0) scale(1.33);
    }
  }

  .as-range-slider__thumb-handle--moving + .as-range-slider__value--moving {
    transform: translate3d(-53%, 4px, 0);
  }
}
