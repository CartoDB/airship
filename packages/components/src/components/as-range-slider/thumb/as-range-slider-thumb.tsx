import { Component, Element, Event, EventEmitter, Listen, Prop } from '@stencil/core';
import { MouseTrack } from '../MouseTrack';

@Component({
  shadow: false,
  styleUrl: './as-range-slider-thumb.scss',
  tag: 'as-range-slider-thumb',
})
export class RangeSliderThumb extends MouseTrack {
  @Prop() public percentage: number;
  @Prop() public value: number;
  @Prop() public valueMin: number;
  @Prop() public valueMax: number;
  @Prop() public disabled: boolean;
  @Prop() public formatValue: (value: number) => void;

  @Event() public thumbMove: EventEmitter<number>;
  @Event() public thumbChangeStart: EventEmitter<void>;
  @Event() public thumbChangeEnd: EventEmitter<void>;
  @Event() public thumbIncrease: EventEmitter<number>;
  @Event() public thumbDecrease: EventEmitter<number>;


  @Element() public element: HTMLElement;
  public railElement: HTMLElement;
  public thumbValue: HTMLElement;
  public railBoundingClientRect: ClientRect | DOMRect;

  public render() {
    const thumbStyles = {
      left: `${this.percentage}%`
    };

    const cssClasses = {
      'as-range-slider__thumb': true,
      'as-range-slider__thumb--disabled': this.disabled
    };

    const cssValueClasses = {
      'as-caption': true,
      'as-font-medium': true,
      'as-range-slider__value': true,
      'as-range-slider__value--disabled': this.disabled,
    };

    return (
      <div role='slider'
        tabindex={this.disabled ? '-1' : '0'}
        aria-valuetext={this._getDisplayValue(this.value)}
        aria-valuenow={this.value}
        aria-valuemin={this.valueMin}
        aria-valuemax={this.valueMax}
        class={cssClasses} style={thumbStyles} data-value={this.value}>
        <div class='as-range-slider__thumb-handle'></div>
        <span class={cssValueClasses}>
          {this._getDisplayValue(this.value)}
        </span>
      </div>);
  }

  @Listen('mousedown')
  @Listen('touchstart')
  public onMouseDown(event: MouseEvent) {
    this.thumbChangeStart.emit();
    this.railElement = document.querySelector('.as-range-slider__rail');

    const thumb = event.target as HTMLElement;
    thumb.classList.add('as-range-slider__thumb-handle--moving');

    this.thumbValue = thumb.parentElement.querySelector('.as-range-slider__value');
    this.thumbValue.classList.add('as-range-slider__value--moving');

    this.railBoundingClientRect = this.railElement.getBoundingClientRect();

    super.handleMouseDown({
      move: (moveEvent) => this._onMove(moveEvent),
      release: () => this._onRelease(thumb)
    });

    thumb.focus();
  }

  @Listen('keydown')
  public onKeyDown(event: KeyboardEvent) {
    if (this.disabled) {
      return;
    }

    const KEY = {
      DOWN: 40,
      LEFT: 37,
      RIGHT: 39,
      UP: 38
    };

    let flag = false;
    switch (event.keyCode) {
      case KEY.DOWN:
      case KEY.LEFT:
        this.thumbDecrease.emit();
        flag = true;
        break;
      case KEY.UP:
      case KEY.RIGHT:
        this.thumbIncrease.emit();
        flag = true;
        break;
      default:
        break;
    }

    if (flag) {
      event.preventDefault();
      event.stopPropagation();
    }
  }

  private _onMove(event: MouseEvent | TouchEvent) {
    this.setCursorTo('grabbing');
    const eventX = event instanceof TouchEvent ? event.changedTouches[0].pageX : event.pageX;
    const barPercentage = (eventX - this.railBoundingClientRect.left) * 100 / this.railElement.offsetWidth;

    if (barPercentage < 0 || barPercentage > 100) {
      return;
    }

    this.thumbMove.emit(barPercentage);
  }

  private _onRelease(thumb: HTMLElement) {
    thumb.classList.remove('as-range-slider__thumb-handle--moving');
    this.thumbValue.classList.remove('as-range-slider__value--moving');

    this.setCursorTo('');

    this.thumbChangeEnd.emit();
  }

  private _getDisplayValue(value: number) {
    return (this.formatValue && this.formatValue(value)) || value;
  }

  private setCursorTo(value) {
    document.body.style.cursor = value;
  }
}

export interface Thumb {
  value: number;
  valueMin: number;
  valueMax: number;
  percentage: number;
}
