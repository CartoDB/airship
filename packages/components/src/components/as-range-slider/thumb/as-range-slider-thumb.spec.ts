import { TestWindow } from '@stencil/core/dist/testing';
import { RangeSliderThumb } from './as-range-slider-thumb';

describe('as-range-slider-thumb', () => {
    it('should build', () => {
        expect(new RangeSliderThumb()).toBeTruthy();
    });

    describe('Rendering', () => {
        let element: HTMLAsRangeSliderThumbElement;
        let testWindow: TestWindow;

        beforeEach(async () => {
            testWindow = new TestWindow();
            element = await testWindow.load({
                components: [RangeSliderThumb],
                html: '<as-range-slider-thumb></as-range-slider-thumb>'
            });
        });

        it('should render a text with its value', async () => {
            element.value = 5;
            element.formatValue = () => `${element.value}€`;

            await testWindow.flush();
            expect(element).toMatchSnapshot();
        });

        it('should have aria attributes', async () => {
            element.value = 5;
            element.valueMin = 0;
            element.valueMax = 10;
            element.percentage = 10;
            element.formatValue = () => `${element.value}€`;
            await testWindow.flush();

            expect(element).toMatchSnapshot();
        });
    });

    describe('Interaction', () => {
        let element: HTMLAsRangeSliderThumbElement;
        let testWindow: TestWindow;

        beforeEach(async () => {
            testWindow = new TestWindow();
            element = await testWindow.load({
                components: [RangeSliderThumb],
                html: '<as-range-slider-thumb></as-range-slider-thumb>'
            });

            this.simulateKeyDownEvent = (targetElement: HTMLElement, code: number) => {
                const event = new (testWindow.window as any).KeyboardEvent('keydown', {
                    bubbles: true,
                    cancelable: true,
                    keyCode: code
                });
                targetElement.dispatchEvent(event);
                return event;
            };

            this.simulateMouseEvent = (targetElement: HTMLElement, type: string, eventData: any) => {
                const data = Object.assign({
                    bubbles: true,
                    cancelable: true
                }, eventData);
                const mouseEvent = new (testWindow.window as any).MouseEvent(type, data);
                targetElement.dispatchEvent(mouseEvent);
                return mouseEvent;
            };
        });

        it('should emit thumbIncrease / thumbDecrease events when using keyboard arrows', async () => {
            const onThumbIncreaseSpy = jest.fn();
            element.addEventListener('thumbIncrease', onThumbIncreaseSpy);

            const onThumbDecreaseSpy = jest.fn();
            element.addEventListener('thumbDecrease', onThumbDecreaseSpy);
            await testWindow.flush();

            const LEFT_KEY = 37;
            this.simulateKeyDownEvent(element, LEFT_KEY);
            expect(onThumbDecreaseSpy).toHaveBeenCalled();

            const RIGHT_KEY = 39;
            this.simulateKeyDownEvent(element, RIGHT_KEY);
            expect(onThumbIncreaseSpy).toHaveBeenCalled();
        });

        it('should emit changeStart event when using mousemove', async () => {
            const onChangeStartSpy = jest.fn();
            element.addEventListener('thumbChangeStart', onChangeStartSpy);
            await testWindow.flush();

            this.simulateMouseEvent(element, 'mousedown');
            expect(onChangeStartSpy).toHaveBeenCalled();
            expect(element.classList.contains('as-range-slider__thumb-handle--moving')).toBe(true);
        });

        // it('should emit thumbMove event when using mouse', async () => {
        //     const onThumbMoveSpy = jest.fn();
        //     element.addEventListener('thumbMove', onThumbMoveSpy);
        //     await testWindow.flush();

        //     this.simulateMouseEvent(element, 'mousedown');
        //     this.simulateMouseEvent(element, 'mousemove');
        //     this.simulateMouseEvent(element, 'mouseup');

        //     expect(onThumbMoveSpy).toHaveBeenCalled();
        // });

    });
});
