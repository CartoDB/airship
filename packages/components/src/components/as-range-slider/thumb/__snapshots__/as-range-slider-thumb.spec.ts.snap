// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-range-slider-thumb Rendering should have aria attributes 1`] = `
<as-range-slider-thumb
  class="hydrated"
>
  
  <div
    aria-valuemax="10"
    aria-valuemin="0"
    aria-valuenow="5"
    aria-valuetext="5€"
    class="as-range-slider__thumb"
    data-value="5"
    role="slider"
    style="left: 10%;"
    tabindex="0"
  >
    <div
      class="as-range-slider__thumb-handle"
    />
    <span
      class="as-caption as-font-medium as-range-slider__value"
    >
      5€
    </span>
  </div>
</as-range-slider-thumb>
`;

exports[`as-range-slider-thumb Rendering should render a text with its value 1`] = `
<as-range-slider-thumb
  class="hydrated"
>
  
  <div
    aria-valuenow="5"
    aria-valuetext="5€"
    class="as-range-slider__thumb"
    data-value="5"
    role="slider"
    tabindex="0"
  >
    <div
      class="as-range-slider__thumb-handle"
    />
    <span
      class="as-caption as-font-medium as-range-slider__value"
    >
      5€
    </span>
  </div>
</as-range-slider-thumb>
`;
